projects/pgai/pgai/vectorizer/embeddings.py

```diff
@@ -414,3 +469,191 @@ async def _encode(self, documents: list[str]) -> list[list[int]]:
+    num_thread: int
+
+    # runtime options
+    num_keep: int
+    seed: int
+    num_predict: int
+    top_k: int
+    top_p: float
+    tfs_z: float
+    typical_p: float
+    repeat_last_n: int
+    temperature: float
+    repeat_penalty: float
+    presence_penalty: float
+    frequency_penalty: float
+    mirostat: int
+    mirostat_tau: float
+    mirostat_eta: float
+    penalize_newline: bool
+    stop: Sequence[str]
+
+
+class Ollama(BaseModel, Embedder):
+    """
+    Embedder that uses Ollama to embed documents into vector representations.
+
+    Attributes:
+        implementation (Literal["ollama"]): The literal identifier for this
+            implementation.
+        model (str): The name of the Ollama model used for embeddings.
+        base_url (str): The base url used to access the Ollama API.
+        truncate (bool): Truncate input longer than the model's context length
+        options (dict): Additional ollama-specific runtime options
+        keep_alive (str): How long to keep the model loaded after the request
+    """
+
+    implementation: Literal["ollama"]
+    model: str
+    base_url: str | None = None
+    truncate: bool = True
+    options: OllamaOptions | None = None
+    keep_alive: str | None = None  # this is only `str` because of the SQL API
+
+    @override
+    async def embed(
+        self, documents: list[str]
+    ) -> Sequence[EmbeddingVector | ChunkEmbeddingError]:
+        """
+        Embeds a list of documents into vectors using Ollama's embeddings API.
+
+        If a request to generate embeddings fails because one or more chunks
+        exceed the model's token limit (and truncate is set to False), every
+        chunk will be retried individually. The returned result will contain a
+        ChunkEmbeddingError in place of an EmbeddingVector for the chunks that
+        exceeded the model's token limit.
+
+        Args:
+            documents (list[str]): A list of documents to be embedded.
+
+        Returns:
+            Sequence[EmbeddingVector | ChunkEmbeddingError]: The embeddings or
+            errors for each document.
+        """
+        await logger.adebug(f"Chunks produced: {len(documents)}")
+        try:
+            return await self._batcher.batch_chunks_and_embed(documents)
+        except ollama.ResponseError as e:
+            if "input length exceeds maximum context length" not in e.error:
+                raise e
+            # Note: We don't attempt tokenizing data sent to Ollama models
+            # because every model has its own tokenizer, and there's no way to
+            # programmatically determine the tokenizer. Without knowing the
+            # token length of a chunk, we can't filter the long ones out (like
+            # we do in the OpenAI integration). Instead, we retry each chunk
+            # individually. It's not ideal, but not too bad because it's
+            # (probably) a local API.
+            context_length = await self._context_length()
```

Comment:
Just to understand, Is the `truncate=true` preventing this? Or we can still fall into this anyway?

Reply:
Yes, `truncate=True` will prevent us getting the error "input length exceeds maximum context length".

Reply:
Worth to add a comment here 🙏 feel free to discard otherwise.